### PR TITLE
[SPARK-25083][SQL] Remove the type erasure hack in data source scan

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

1. Add function `inputBatchRDDs` and `inputRowRDDs` in `ColumnarBatchScan`.
2. Rewrite the 2 functions in physical nodes which extend `ColumnarBatchScan`.
3. Remove hack alert in `InMemoryTableScanExec` and `FileScanRDD`.
https://github.com/apache/spark/blob/ab33028957443189efc4106afd9d65dddf8f9c98/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala#L108-L109
https://github.com/apache/spark/blob/ab33028957443189efc4106afd9d65dddf8f9c98/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala#L219
4. Remove TODOs in `DataSourceRDD`.
https://github.com/apache/spark/blob/ab33028957443189efc4106afd9d65dddf8f9c98/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala#L28-L29
https://github.com/apache/spark/blob/ab33028957443189efc4106afd9d65dddf8f9c98/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala#L75

## How was this patch tested?

Existing UT.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala [hunk 1]
```diff
@@ -40,6 +42,29 @@ private[sql] trait ColumnarBatchScan extends CodegenSupport {
...
+   * Returns all the RDDs of InternalRow which generates the input rows.
+   */
+  def inputRowRDDs(): Seq[RDD[InternalRow]]
+
+  /**
+   * Get input RDD depends on supportsBatch.
+   */
+  final def getInputRDDs(): Seq[RDD[InternalRow]] = {
+    if (supportsBatch) {
+      inputBatchRDDs().asInstanceOf[Seq[RDD[InternalRow]]]
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: Here maybe the last explicitly erasure hack left, please check whether is it acceptable or not.

**Comment by xuanyuanking** [comment]
cc @cloud-fan and @rdblue have a look when you have time. If this PR doesn't coincide with your expect, I'll close this soon. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -307,7 +308,7 @@ case class FileSourceScanExec(
     withSelectedBucketsCount
   }
 
-  private lazy val inputRDD: RDD[InternalRow] = {
+  private lazy val inputRDD: RDD[Object] = {
```
[the remark this answers is not in the document]
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Thanks Sean! Addressed in 7e88599.

**Comment by rdblue** [comment]
@xuanyuanking, while this does remove the hack, it doesn't address the underlying problem. The problem is that there is a single RDD, which may contain InternalRow or may contain ColumnarBatch. Generated code knows how to differentiate between the two and use the RDD contents correctly.

While this is an improvement because it uses the actual type of records in the RDD, the work that needs to be done is to update the columnar case so that it does return an `RDD[InternalRow]` for anyone that accesses data using that RDD, and then update the generated code to detect a data source RDD and access the underlying `RDD[ColumnarBatch]`.

Here's some pseudo-code to demonstrate what I mean. The current code does something like this with a cast. Your change wouldn't fix the need to cast to `RDD[ColumnarBatch]`:
```scala
def doExecute(rdd: DataSourceRDD[InternalRow]) { // with your change, DataSourceRDD[_]
  if (rdd.isColumnar) {
    doExecuteColumnarBatch(rdd.asInstanceOf[RDD[ColumnarBatch]])
  } else {
    doExecuteRows(rdd)
  }
}
```

I think that should be changed to something like this which is type safe:
```scala
def doExecute(rdd: DataSourceRDD[InternalRow]) {
  if (rdd.isColumnar) {
    doExecuteColumnarBatch(rdd.getColumnBatchRDD)
  } else {
    doExecuteRows(rdd)
  }
}
```

**Comment by xuanyuanking** [comment]
@rdblue Great thanks for your detailed reply! I think I can understand what should be done in this work by your pseudo-code. Maybe its the same problem with the last hack here: https://github.com/apache/spark/pull/22222#discussion_r212784374. I'll keep looking into this.

**Comment by xuanyuanking** [comment]
@cloud-fan Thanks for your reply Wenchen, I'm trying to achieve this in this commit, please take a look, thanks.

**Comment by cloud-fan** [comment]
can we do it for data source v2 first? It seems hard to fix the file source, as its reader function may lie about the return type.

Let's see what's the simplest fix to remove the hack for data source v2.

**Comment by xuanyuanking** [comment]
Got it, I'll revert the changes in file source in this commit, thanks for your reply.

**Pull request closed without merging**